src/languages/handlebars.js

```diff
@@ -8,37 +8,163 @@ Category: template
 */
 
 export default function(hljs) {
-  var BUILT_INS = {'builtin-name': 'each in with if else unless bindattr action collection debugger log outlet template unbound view yield lookup'};
+  const BUILT_INS = {
+    'builtin-name': [
+      'action',
+      'bindattr',
+      'collection',
+      'component',
+      'concat',
+      'debugger',
+      'each',
+      'each-in',
+      'get',
+      'hash',
+      'if',
+      'in',
+      'input',
+      'link-to',
+      'loc',
+      'log',
+      'lookup',
+      'mut',
+      'outlet',
+      'partial',
+      'query-params',
+      'render',
+      'template',
+      'textarea',
+      'unbound',
+      'unless',
+      'view',
+      'with',
+      'yield'
+    ].join(" ")
+  };
 
-  var IDENTIFIER_PLAIN_OR_QUOTED = {
-    begin: /".*?"|'.*?'|\[.*?\]|\w+/
+  const LITERALS = {
+    literal: [
+      'true',
+      'false',
+      'undefined',
+      'null'
+    ].join(" ")
   };
 
-  var EXPRESSION_OR_HELPER_CALL = hljs.inherit(IDENTIFIER_PLAIN_OR_QUOTED, {
-    keywords: BUILT_INS,
+  // as defined in https://handlebarsjs.com/guide/expressions.html#literal-segments
+  // this regex matches literal segments like ' abc ' or [ abc ] as well as helpers and paths
+  // like a/b, ./abc/cde, and abc.bcd
+  const IDENFIFIER_REGEX = /(".*?"|'.*?'|\[.*?\]|[^\s!"#%&'()*+,.\/;<=>@\[\\\]^`{|}~]+|\.|\/)+/;
+
+  // identifier followed by a equal-sign (without the equal sign)
+  const HASH_PARAM_REGEX = /(".*?"|'.*?'|\[.*?\]|[^\s!"#%&'()*+,.\/;<=>@\[\\\]^`{|}~]+)(?==)/;
```

Comment:
A little duplication here, could we pull out the "IDENT" part and use it for both expressions?  Look at the new `regex.concat` helper if you want a nice abstraction.

Reply:
I actually found out the the `HASH_PARAM_REGEX` is wrong.

```
{{{helper [a}}a]=x}}}
```

is valid (`a}}a` is the key), but 

```
{{{helper 'a}}a'=x}}}
```
and 

```
{{{helper "a}}a"'=x}}}
```
are not. I don't know if we add this in the future.
I have tried a refactoring, but since the two regexes have more differences now, it has become a little more complex. Hopefully clearer, though.

Reply:
I liked your refactoring, a bit more than I had in mind, but it's clean and clear.